apps/demo-ui/openai-proxy/py_proxy/token_proxy.py

```diff
@@ -0,0 +1,118 @@
+from gevent import monkey
+
+# ruff: noqa: E402
+monkey.patch_all()  # Must be before other imports
+
+import os
+import sys
+import logging
+import requests
+import warnings
+import urllib3
+
+from gevent.socket import socket
+from gevent.pywsgi import WSGIServer
+from flask import abort, current_app, Flask, make_response, redirect, request, url_for
+from flask_cors import CORS
+
+warnings.filterwarnings("ignore", category=urllib3.exceptions.InsecureRequestWarning)
+
+# This needs to be global to make decorators work...
+app = Flask("token_proxy", static_folder=None)
+
+UID = 1000  # This is the Docker default
+UI_PORT = 3000
+METHODS = ["GET", "POST", "PUT", "DELETE", "HEAD", "OPTIONS"]
+
+BAD_HEADERS = {
+    "content-encoding",
+    "content-length",
+    "transfer-encoding",
+    "connection",
+}
+
+
+@app.route("/tok/<tok>")
+def proxy_token(tok="", arg=""):
+    token = current_app.config["token"]
+    if tok != token:
+        logging.warning(f"bad token {tok}")
+        abort(403)
+
+    resp = redirect(url_for("proxy_plain"))
+    resp.set_cookie("token", token)
+    return resp
+
+
+@app.route("/", methods=METHODS)
+@app.route("/<path:arg>", methods=METHODS)
+def proxy_plain(arg=""):
+    tok = request.cookies.get("token")
+    if not tok:
+        logging.warning("missing token")
+        abort(403)
+    token = current_app.config["token"]
```

Comment:
1.  These are session cookies that don't outlive the browser session.  New values will replace old.  If users always start with the token URL, they should be fine.  This should be the same behavior as what we're doing with Jupyter.

2.  It's as secure as the Jupyter one.  It's a bearer token essentially.  Anyone seeing the Docker logs has access to it.  It's transmitted over SSL and it's specific to the domain/hostname.  Since it's a session cookie, it shouldn't get stored to disk on the user's end.  It would be hard to steal from the browser, and stealing from logs means you already have access.